ec2/ec2test/server.go

```diff
@@ -2163,6 +2172,249 @@ func (srv *Server) unassignPrivateIP(w http.ResponseWriter, req *http.Request, r
+			volume.AvailZone = v
+		case "SnapshotId":
+			volume.SnapshotId = vals[0]
+		case "VolumeType":
+			volume.VolumeType = vals[0]
+		case "Size":
+			volume.Size = atoi(vals[0])
+		case "Iops":
+			volume.IOPS = int64(atoi(vals[0]))
+		case "Encrypted":
+			val, err := strconv.ParseBool(vals[0])
+			if err != nil {
+				fatalf(400, "InvalidParameterValue", "bad flag %s: %s", attr, vals[0])
+			}
+			volume.Encrypted = val
+		default:
+			fatalf(400, "InvalidParameterValue", "unknown field %s: %s", attr, vals[0])
+		}
+	}
+	volume.Id = fmt.Sprintf("vol-%d", srv.volumeId.next())
+	volume.Status = "available"
+	volume.CreateTime = time.Now().Format(time.RFC3339)
+	if volume.VolumeType == "" {
+		volume.VolumeType = "magnetic"
+	}
+	if volume.Size == 0 {
+		volume.Size = 1
+	}
+	return volume
+}
+
+func (srv *Server) deleteVolume(w http.ResponseWriter, req *http.Request, reqId string) interface{} {
+	v := srv.volume(req.Form.Get("VolumeId"))
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+
+	delete(srv.volumes, v.Id)
+	return &ec2.SimpleResp{
+		XMLName:   xml.Name{defaultXMLName, "DeleteVolumeResponse"},
+		RequestId: reqId,
+		Return:    true,
+	}
+}
+
+func (srv *Server) volume(id string) *volume {
+	if id == "" {
+		fatalf(400, "MissingParameter", "missing volumeId")
+	}
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+	v, found := srv.volumes[id]
+	if !found {
+		fatalf(400, "InvalidVolumeID.NotFound", "Volume %s not found", id)
+	}
+	return v
+}
+
+func (srv *Server) describeVolumes(w http.ResponseWriter, req *http.Request, reqId string) interface{} {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+
+	idMap := parseIDs(req.Form, "VolumeId.")
+	f := newFilter(req.Form)
+	var resp struct {
+		XMLName xml.Name
+		ec2.VolumesResp
+	}
+	resp.XMLName = xml.Name{defaultXMLName, "DescribeVolumesResponse"}
+	resp.RequestId = reqId
+	for _, v := range srv.volumes {
+		ok, err := f.ok(v)
+		_, known := idMap[v.Id]
+		if ok && (len(idMap) == 0 || known) {
+			resp.Volumes = append(resp.Volumes, v.Volume)
+		} else if err != nil {
+			fatalf(400, "InvalidParameterValue", "describe Volumes: %v", err)
+		}
+	}
+	return &resp
+}
+
+type volumeAttachment struct {
+	ec2.VolumeAttachment
+}
+
+func (srv *Server) attachVolume(w http.ResponseWriter, req *http.Request, reqId string) interface{} {
+	if req.Form.Get("VolumeId") == "" {
+		fatalf(400, "MissingParameter", "missing volume id")
+	}
+	if req.Form.Get("InstanceId") == "" {
+		fatalf(400, "MissingParameter", "missing instance id")
+	}
+	if req.Form.Get("Device") == "" {
+		fatalf(400, "MissingParameter", "missing device")
+	}
+	ec2VolAttachment := srv.parseVolumeAttachment(req)
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+	va := &volumeAttachment{ec2VolAttachment}
+	srv.volumeAttachments[va.VolumeId] = va
+	var resp struct {
+		XMLName xml.Name
+		ec2.VolumeAttachmentResp
+	}
+	resp.XMLName = xml.Name{defaultXMLName, "AttachVolumeResponse"}
+	resp.RequestId = reqId
+	resp.VolumeId = va.VolumeId
+	resp.InstanceId = va.InstanceId
```

Comment:
Instance id validation is done in parseVolume attachment. Added availability zone check.